Reorder requirements file decoding

matthewhughes934 opened the pull request:
This changes the decoding process to be more in line with what was
previously documented. The new process is outlined in the updated docs.

The `auto_decode` function was removed and all decoding logic moved to
the `pip._internal.req.req_file` module because:

* This function was only ever used to decode requirements file
* It was never really a generic 'util' function, it was always tied to
  the idiosyncrasies of decoding requirements files.
* The module lived under `_internal` so I felt comfortable removing it

A warning was added when we _do_ fallback to using the locale defined
encoding to encourage users to move to an explicit encoding definition
via a coding style comment.

This fixes two existing bugs. Firstly, when:

* a requirements file is encoded as UTF-8, and
* some bytes in the file are incompatible with the system locale

Previously, assuming no BOM or PEP-263 style comment, we would default
to using the encoding from the system locale, which would then fail (see
issue #12771)

Secondly, when decoding a file starting with a UTF-32 little endian Byte
Order Marker. Previously this would always fail since
`codecs.BOM_UTF32_LE` is `codecs.BOM_UTF16_LE` followed by two null
bytes, and because of the ordering of the list of BOMs we the UTF-16
case would be run first and match the file prefix so we would
incorrectly deduce that the file was UTF-16 little endian encoded. I
can't imagine this is a popular encoding for a requirements file.

Fixes: #12771

matthewhughes934 commented on news/9a3e9584-3fd4-4840-916b-414c164f9c28.trivial.rst:
Ah, this probably needs a proper news entry

ichard26 replied:
Yes, please.

matthewhughes934 commented on src/pip/_internal/req/req_file.py:
```diff
@@ -545,7 +546,23 @@ def get_file_content(url: str, session: "PipSession") -> Tuple[str, str]:
+            UnicodeWarning,
+            stacklevel=2,
+        )
+        content = raw_content.decode(fallback_encoding)
```
> It would be ideal to include filename or filepath of the requirements file.

Agreed, [cae26c0](https://github.com/pypa/pip/pull/12795/commits/cae26c0c1f50a598e31cf715eb00bc79505188fb)

matthewhughes934 commented on src/pip/_internal/req/req_file.py:
```diff
@@ -545,7 +546,23 @@ def get_file_content(url: str, session: "PipSession") -> Tuple[str, str]:
+            f"unable to decode data with {exc.encoding}, falling back to {fallback_encoding}",  #  noqa: E501
+            UnicodeWarning,
+            stacklevel=2,
+        )
```
> The problem with using `warnings.warn` is that its presentation format is inappropriately technical. `logger.warning` should be used instead.

I think I just went with this because I knew `UnicodeWarning` was a thing, happy to go with logging [cae26c0](https://github.com/pypa/pip/pull/12795/commits/cae26c0c1f50a598e31cf715eb00bc79505188fb)

uranusjr commented on src/pip/_internal/req/req_file.py:
```diff
@@ -545,7 +545,25 @@ def get_file_content(url: str, session: "PipSession") -> Tuple[str, str]:
+            exc.encoding,
+            fallback_encoding,
+        )
+        content = raw_content.decode(fallback_encoding)
```
It may be a good idea to use `error="backslashreplace"` here. Most of the time, the offending bytes would just be a part of a comment anyway and would not make a difference.

matthewhughes934 replied:
> It may be a good idea to use `error="backslashreplace"` here. Most of the time, the offending bytes would just be a part of a comment anyway and would not make a difference.

I've been hesitating with this a bit, specifically I'm wondering if this could be abused for nefarious purposes where the contents of the file you 'see' (not well defined, since this is the case where the data won't fully decode) isn't the same contents that pip will process. Though I'm having a hard time finding a vulnerable use-case (e.g injecting an extra element or a adding a `.` to a domain name in a requirement URL)

ichard26 replied:
Generally, we're trying to enforce compliance with the documentation, so I don't think we should introduce another hack to support technically mal-encoded requirement files.

sbidoul commented:
Hmm, since the documentation says it's utf-8 unless there is a PEP-263 style comment, shouldn't we rather decode as utf8 is there is no such comment, and if that fails, fallback to the current `locale.getpreferredencoding(False) or sys.getdefaultencoding()` with a deprecation warning recommending to add the encoding comment?

That way we have a (more or less) non-breaking path to compliance with the docs?

Also, I'd put all that in auto_decode, with a docstring comment that the function is meant for requirements.txt decoding as per the docs.

matthewhughes934 commented:
> Hmm, since the documentation says it's utf-8 unless there is a PEP-263 style comment, shouldn't we rather decode as utf8 is there is no such comment, and if that fails, fallback to the current `locale.getpreferredencoding(False) or sys.getdefaultencoding()` with a deprecation warning recommending to add the encoding comment?
> 
> That way we have a (more or less) non-breaking path to compliance with the docs?
> 
> Also, I'd put all that in auto_decode, with a docstring comment that the function is meant for requirements.txt decoding as per the docs.

This sounds reasonable, though I think this change would need to be made in `auto_deocde` rather than where I made it. But now I am wondering if `auto_decode` needs to live where it does: it's only ever used for decoding requirements files, so maybe it can just live in `req_file`

sbidoul commented:
Sounds good. I've removed from the 24.3 milestone. Feel free to ping me when you get back to this.

matthewhughes934 commented:
> Sounds good. I've removed from the 24.3 milestone. Feel free to ping me when you get back to this.

:+1: I've updated the change. It was basically a re-do so I just stomped my previous commits.

ichard26 commented on docs/html/reference/requirements-file-format.md:
```diff
@@ -56,9 +56,15 @@ examples of all these forms, see {ref}`pip install Examples`.
-Requirements files are `utf-8` encoding by default and also support
-{pep}`263` style comments to change the encoding (i.e.
-`# -*- coding: <encoding name> -*-`).
+It is simplest to encode your requirements files with UTF-8.
+The process for decoding requirements files is:
+
+- Check for any Byte Order Mark at the start of the file and if found use
+  the corresponding encoding to decode the file.
+- Check for any {pep}`263` style comment (e.g. `# -*- coding: <encoding name> -*-`)
+  and if found decode with the given encoding.
+- Try and decode with UTF-8, and if that fails,
+- fallback to trying to decode using the locale defined encoding.
```
This is the specification for requirement files, we should be tighter. 
````md
The default encoding for requirement files is `UTF-8` unless a different encoding
is specified using a {pep}`263` style comment (e.g. `# -*- coding: <encoding name> -*-`). 

```{warning}
pip will fallback to the locale defined encoding if `UTF-8` decoding fails. This is a quirk
of pip's parser. This behaviour is *deprecated* and should not be relied upon.
```
````

While we're here, should we expand the specification to formally allow BOMs?

matthewhughes934 replied:
> While we're here, should we expand the specification to formally allow BOMs?

How much do we want to encourage this usage, do you have an idea for how much they are relied on? Specifically, I'm wondering if we can just say: UTF-8 by default, PEP-263 style comments supported, other bits deprecated (if someone is relying on locale specific or BOM, they should switch to PEP-263 comment)

pfmoore replied:
I would prefer to *not* allow BOMs, as @matthewhughes934 says. Ideally, as it's a backward incompatible change, we would initially issue a deprecation warning if a BOM is found, but given that the use of a BOM is undocumented (in the [user docs](https://pip.pypa.io/en/stable/reference/requirements-file-format/#encoding)), I'd be fine with simply removing that support if that's easier.

However, PEP 263 states that the UTF8 signature `'\xef\xbb\xbf'` should signal UTF8, so that one (which is treated as a BOM in `auto_decode`) should stay.

It's worth noting that UTF16 and UTF32 encodings can't be handled with a PEP 263 style comment, as those encodings are not ASCII supersets. But I think it's fine to stop supporting them.

notatallshaw replied:
> It's worth noting that UTF16 and UTF32 encodings can't be handled with a PEP 263 style comment, as those encodings are not ASCII supersets. But I think it's fine to stop supporting them.

I'm not following this issue closely, but in case it's helpful, I'll note that uv initially only supported UTF8 but then found they had user issues not supporting UTF16: https://github.com/astral-sh/uv/pull/2283

I think because redirecting from stdout to a file on Windows would end up with a UTF16 file, e.g. `pip freeze > requirements.txt`.

ichard26 replied:
I'm not really in the mood to break our Windows users over requirements file decoding, honestly. I'd prefer documenting BOMs as supported as I don't like to have implementation-defined behaviour here, but I also recognize that would _promote_ their proliferation so leaving this as implementation-defined behaviour is a fair compromise. Using the old docs text or my suggestion (without any changes) is fine. 

Thanks @notatallshaw for letting us know that UTF-16 can be pretty common on Windows.

matthewhughes934 replied:
> I'm not really in the mood to break our Windows users over requirements file decoding, honestly. I'd prefer documenting BOMs as supported as I don't like to have implementation-defined behaviour here, but I also recognize that would _promote_ their proliferation so leaving this as implementation-defined behaviour is a fair compromise. Using the old docs text or my suggestion (without any changes) is fine.

:+1: I like your more concise docs, and the warning, so went with that 5c49381e4

pfmoore replied:
> I think because redirecting from stdout to a file on Windows would end up with a UTF16 file

In Windows 11, cmd seems to use UTF8[^1], as does Powershell Core. Windows Powershell does use UTF16, though, so I guess there's a proportion of users who will encounter this (I don't have a feel for the relative usage of the 3 shells).

Given this, +1 on retaining the existing support but not documenting it.

[^1]: This may depend on the console codepage, I'm not sure when UTF8 became the default.

ichard26 replied:
:+1: I added a code comment explaining why we still support BOMs: https://github.com/pypa/pip/pull/12795/commits/f1dbf496677b950680ceed74a76cccaf0fe6cf17

ichard26 commented on src/pip/_internal/req/req_file.py:
```diff
@@ -563,7 +565,56 @@ def get_file_content(url: str, session: "PipSession") -> Tuple[str, str]:
+
+
+def _decode_req_file(data: bytes, url: str) -> str:
+    # order of BOMS is important: codecs.BOM_UTF16_LE is a prefix of codecs.BOM_UTF32_LE
+    # so data.startswith(BOM_UTF16_LE) would be true for UTF32_LE data
```
This comment should be moved up within the BOMS list.

matthewhughes934 replied:
> This comment should be moved up within the BOMS list.

2f3c2681d51ca1165befab472c1418ac1e19a1bd I also did some extra tidying with afac4fb2c42b6a674abe494f6f0b138c09b40b46

ichard26 commented on src/pip/_internal/req/req_file.py:
```diff
@@ -563,7 +565,56 @@ def get_file_content(url: str, session: "PipSession") -> Tuple[str, str]:
     return url, content
+
+
+BOMS: List[Tuple[bytes, str]] = [
+    (codecs.BOM_UTF8, "utf-8"),
+    (codecs.BOM_UTF32, "utf-32"),
+    (codecs.BOM_UTF32_BE, "utf-32-be"),
+    (codecs.BOM_UTF32_LE, "utf-32-le"),
+    (codecs.BOM_UTF16, "utf-16"),
+    (codecs.BOM_UTF16_BE, "utf-16-be"),
+    (codecs.BOM_UTF16_LE, "utf-16-le"),
+]
+
+ENCODING_RE = re.compile(rb"coding[:=]\s*([-\w.]+)")
+DEFAULT_ENCODING = "utf-8"
```
These should be moved to near the top of the module where any other constant would live.

matthewhughes934 replied:
> These should be moved to near the top of the module where any other constant would live.

Done with 2f3c2681d51ca1165befab472c1418ac1e19a1bd

ichard26 commented on src/pip/_internal/req/req_file.py:
```diff
@@ -563,7 +565,56 @@ def get_file_content(url: str, session: "PipSession") -> Tuple[str, str]:
     # Assume this is a bare path.
     try:
         with open(url, "rb") as f:
-            content = auto_decode(f.read())
+            raw_content = f.read()
     except OSError as exc:
         raise InstallationError(f"Could not open requirements file: {exc}")
+
+    content = _decode_req_file(raw_content, url)
```
Hmm, if the decoding with the locale encoding also fails, the exception won't be caught. Any objections to calling `_decode_req_file` within in the try-except?

matthewhughes934 replied:
> Hmm, if the decoding with the locale encoding also fails, the exception won't be caught. Any objections to calling `_decode_req_file` within in the try-except?

No objection to adding it there. My original assumption, and the reason I moved it out, was the catching of `OSError` was only really relevant for reading the content, and not the actual decoding.

ichard26 commented on tests/unit/test_req_file.py:
```diff
@@ -942,3 +943,116 @@ def test_install_requirements_with_options(
+        assert str(reqs[0].specifier) == "<=24.0"
+
+    @pytest.mark.parametrize("encoding", ["utf-8", "gbk"])
+    def test_erorrs_on_non_decodable_data(
```
```suggestion
    def test_errors_on_non_decodable_data(
```
:slightly_smiling_face:

matthewhughes934 replied:
:upside_down_face: fd2801e76bf134d1c2380a91c6cd8e2523220a84

ichard26 reviewed (changes requested):
Looks great so far! I love your tests. They're extensive and so easy to read!

ichard26 commented:
Oh and sorry for taking so long to re-review your PR. It fell off our radar, and then we haven't had the time to review it honestly.

ichard26 reviewed (approved):
We could make the fallback to the locale/system encoding raise a deprecation warning instead of a plain warning, but I'm honestly not very interested in going through a deprecation cycle over this. If other maintainers feel differently, I have no issues with adding one.

Thank you again for your patience! Good work!

matthewhughes934 commented:
Latest change was just the result of me rebasing on `main` and squashing all the fixup commits together

matthewhughes934 commented:
There's a test failure https://github.com/pypa/pip/actions/runs/12528829883/job/34943820232?pr=12795, error:

```
FAILED tests/functional/test_install.py::test_vcs_url_urlquote_normalization - pip._internal.exceptions.InstallationSubprocessError: bzr checkout --lightweight --quiet http://bazaar.launchpad.net/%7Edjango-wikiapp/django-wikiapp/release-0.1 /tmp/pytest-of-runner/pytest-1/popen-gw1/test_vcs_url_urlquote_normaliz0/cache/release-0.1 exited with 3
```

It looks like `bzr checkout` failed with an error? Maybe some intermittent/network issue: it ran ok for all other Python versions+OS (I can't see the button to re-run jobs, I assume due to lack of permissions)

pfmoore commented:
I hit "rerun failed jobs". Let's see if that fixes it.

ichard26 commented:
@pfmoore do you have any objections with the current requirements file specification. I've copied it below for posterity. 

The default encoding for requirement files is `UTF-8` unless a different encoding is specified using a [PEP 263](https://peps.python.org/pep-0263/) style comment (e.g. `# -*- coding: <encoding name> -*-`).

> [!warning]
> pip will fallback to the locale defined encoding if `UTF-8` decoding fails. This is a quirk of pip's parser. This behaviour is *deprecated* and should not be relied upon.

---

In addition, would you like to formally deprecate the fallback to the locale/system encoding? I'm not particularly interested in hard deprecating the fallback and dealing with the churn, especially since it'll be probably be encountered by inexperienced users who are simply using their OS text editor or whatever.

pfmoore commented:
Honestly, I don't care. As long as we don't break what the *current* documentation says:

> Requirements files are utf-8 encoding by default and also support [PEP 263](https://peps.python.org/pep-0263/) style comments to change the encoding (i.e. # -*- coding: <encoding name> -*-).

then we can do whatever we like. I personally would rather we didn't document any more of the current behaviour, as that just constrains us in the future. And I'm not particularly keen on adding a warning that essentially says "we currently do X, but you shouldn't rely on that" when people shouldn't have been relying on it anyway, as it was undocumented. But it's not something I'm going to spend any time debating - if you want to add it, fair enough.

sbidoul commented:
This looks like this is good to go. Any particular reason you postponed it to 25.1 @ichard26 ?

ichard26 commented:
Nothing in particular. I just wasn't sure if everyone was on board with the changes and I didn't want to merge something that was still under active discussion :)

I'll take a look tonight and merge if you're happy with this.

ichard26 commented:
> I personally would rather we didn't document any more of the current behaviour, as that just constrains us in the future.

I don't think that's true in this case? I sympathise that having anything in the documentation lends it credibility and visibility that an undocumented implementation-only detail does not have, but in this case, it's documented explicitly as _deprecated behaviour that should not be relied upon_. For better or worse, while we are moving towards formal specifications, pip's behaviour is still a de facto standard, so it's important to acknowledge the parts where we diverge from the specification IMO.

I haven't spent the past many years writing, reviewing, or implementing standards though, so I'll remove that warning in the documentation. 

> And I'm not particularly keen on adding a warning that essentially says "we currently do X, but you shouldn't rely on that" when people shouldn't have been relying on it anyway, as it was undocumented.

Well, originally the suggestion was to issue a deprecation warning and remove the fallback outright. I guess we're on the same page it's not really worth it. Similar to the documentation callout, I believe it's valuable to warn on standards violations to nudge towards standard compliance.
